### ISSUE TITLE
Request for EntityCallback Support in Quarkus

### ISSUE THREAD

**Issue by nitinty** [quarkusio/quarkus#44066]
### Description

**Summary:** I would like to propose the addition of EntityCallback support in Quarkus, similar to the functionality provided by Spring. Specifically, I am looking for the following callbacks:

**AfterConvertCallback:** Invoked after an entity is converted from its database representation to its domain representation.

**BeforeConvertCallback:** Triggered before an entity is converted to its database representation.

**AfterSaveCallback:** Occurs after an entity is saved to the database.

**BeforeSaveCallback:** Invoked before an entity is saved to the database.

**Use Cases:** Implementing these callbacks would allow developers to execute custom logic at different stages of the entity lifecycle, enhancing flexibility and maintainability.

### Implementation ideas

I’m adding a rough implementation idea here for adding EntityCallback support in Quarkus MongoDB Panache. Please feel free to update or suggest any better approaches!

**Implementation Ideas for EntityCallback Support in Quarkus MongoDB Panache**

**1. Define Callback Interfaces:**

Create interfaces for each callback (e.g., BeforeConvertCallback, AfterConvertCallback, BeforeSaveCallback, AfterSaveCallback).
Each interface can have a method that accepts the entity type as a parameter.

```java

public interface BeforeConvertCallback<T> {
    void beforeConvert(T entity);
}

public interface AfterConvertCallback<T> {
    void afterConvert(T entity);
}

public interface BeforeSaveCallback<T> {
    void beforeSave(T entity);
}

public interface AfterSaveCallback<T> {
    void afterSave(T entity);
}

```

**2. Implement a Callback Registry:**

Create a registry to hold all registered callbacks. This could be done using a Map<Class<?>, List<Callback>> to group callbacks by entity type.

```java 

@Singleton
public class CallbackRegistry {
    private final Map<Class<?>, List<Object>> callbacks = new HashMap<>();

    public <T> void registerCallback(Class<T> entityClass, Object callback) {
        callbacks.computeIfAbsent(entityClass, k -> new ArrayList<>()).add(callback);
    }

    public <T> List<Object> getCallbacks(Class<T> entityClass) {
        return callbacks.getOrDefault(entityClass, Collections.emptyList());
    }
}

```

**3. Enhance the Panache Repository:**

Enhance the PanacheMongoRepository / ReactivePanacheMongoRepository (MongoOperations / ReactiveMongoOperations)

```java

    public void persist(Object entity) {
        invokeCallbacks(entity, BeforeSaveCallback.class);
        MongoCollection collection = mongoCollection(entity);
        persist(collection, entity);
        invokeCallbacks(entity, AfterSaveCallback.class);
    }


    private <T> void invokeCallbacks(T entity, Class<? extends Callback<T>> callbackType) {
        List<Object> callbacks = callbackRegistry.getCallbacks(entity.getClass());
        for (Object callback : callbacks) {
            if (callbackType.isInstance(callback)) {
                ((Callback<T>) callback).invoke(entity);
            }
        }
    }

```

**4. Register Callbacks:**

Allow users to register callbacks using CDI (Contexts and Dependency Injection) annotations, such as @Singleton or @ApplicationScoped.

```java

@Singleton
public class MyEntityCallbacks implements BeforeSaveCallback<MyEntity>, AfterSaveCallback<MyEntity> {

    @Inject
    CallbackRegistry callbackRegistry;

    public MyEntityCallbacks() {
        callbackRegistry.registerCallback(MyEntity.class, this);
    }

    @Override
    public void beforeSave(MyEntity entity) {
        // Custom logic before saving
    }

    @Override
    public void afterSave(MyEntity entity) {
        // Custom logic after saving
    }
}

```

**5. Testing**

Implement component test cases to verify that the callbacks are invoked correctly during the entity lifecycle.

**Comment by nitinty** [comment]
I have opened this issue for #43989 discussion.

**Comment by FroMage** [comment]
I'd rather follow the same semantics as JPA lifecycle listeners, using annotations: https://jakarta.ee/specifications/persistence/3.2/jakarta-persistence-spec-3.2#lifecycle-callback-methods

I could not find any similar lifecyle listener support in Jakarta Data, though perhaps @gavinking can contradict me.

**Comment by nitinty** [comment]
You can, but the approach I mentioned will give developers more freedom to separate any logic from the entity class, keeping it clean and focused. Additionally, it allows custom logic to be placed in a separate class for better organization.

**Comment by nitinty** [comment]
> I'd rather follow the same semantics as JPA lifecycle listeners, using annotations: https://jakarta.ee/specifications/persistence/3.2/jakarta-persistence-spec-3.2#lifecycle-callback-methods
> 
> I could not find any similar lifecyle listener support in Jakarta Data, though perhaps [@gavinking](https://github.com/gavinking) can contradict me.

You can, but the approach I mentioned will give developers more freedom to separate any logic from the entity class, keeping it clean and focused. Additionally, it allows custom logic to be placed in a separate class for better organization.

**Comment by gavinking** [comment]
> I could not find any similar lifecyle listener support in Jakarta Data, though perhaps @gavinking can contradict me.


In *our implementation* of Jakarta Data, you can hook into the,  `@PostPersist`, `@PostUpdate`, and `@PostRemove` events, which are triggered by the underlying stateless session.

My proposal for `EntityAgent` includes a set of events that properly model the stateless operations of a stateless session.

<https://github.com/jakartaee/persistence/pull/675>

There is not yet any firm plan to add Jakarta Data-specific events.

**Comment by gavinking** [comment]
Of course, if you want something higher-level and better-integrated with CDI, then what you are looking for is integration with the CDI event bus. (Much netter than inventing a whole new eventing infrastructure as proposed above.)

It should be quite straightforward to write a JPA `EventListener` which raises CDI events. And then CDI components can received them via `@Observes`.

I'm sure I've displayed an example implementation of such an `EventListener` before.

**Comment by FroMage** [comment]
That was fast 😅

**Comment by nitinty** [comment]
@loicmathieu  @FroMage 
Are we considering the request to add EntityCallback support in Quarkus MongoDB Panache?

**Comment by loicmathieu** [comment]
I think if this would be added to Jakarta Data, it would be a good idea to implement something similar.

But for the moment, Panache is not tied to Jakarta Data, and I'm not sure we want to wait for it to be available, but we can still implement the same and then move to the new event classes when they exist.

And of course, we would need that to be implemented for both Hibernate with Panache and MongoDB with Panache.

**Comment by pip42** [comment]
Any chance this is still being considered, or is there another option to achieve pre/post-insert modifications?

**Comment by FroMage** [comment]
This should be doable in the future if/when we support Mongo via Hibernate ORM.

Until then, someone would have to either provide a PR to support them for Mongo/Panache 1, or use the workaround of overriding your `PanacheEntity.persist()` method to invoke any pre/post save hook. That one's trivial. The convert one less so.
